## Подготовка данных

In [ ]:
import pandas as pd
import numpy as np
import warnings
warnings.simplefilter("ignore")

from sklearn.preprocessing import StandardScaler 
from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_percentage_error

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyRegressor
from lightgbm import LGBMRegressor

In [ ]:
# просмотр, где находится каталог с файлами на COLAB
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# получение доступа к каталогу и уточнение названия папок
import os
os.listdir('/content/drive/My Drive/Colab Notebooks/Яндекс/Проект 9 Определение стоимости автомобилей/ДАННЫЕ')

['autos.csv']

In [ ]:
# загрузка данных
df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Яндекс/Проект 9 Определение стоимости автомобилей/ДАННЫЕ/autos.csv')

In [ ]:
#df = pd.read_csv('/datasets/autos.csv')

In [ ]:
df

DateCrawled  Price  VehicleType  RegistrationYear Gearbox  \
0       2016-03-24 11:52:17    480          NaN              1993  manual   
1       2016-03-24 10:58:45  18300        coupe              2011  manual   
2       2016-03-14 12:52:21   9800          suv              2004    auto   
3       2016-03-17 16:54:04   1500        small              2001  manual   
4       2016-03-31 17:25:20   3600        small              2008  manual   
...                     ...    ...          ...               ...     ...   
354364  2016-03-21 09:50:58      0          NaN              2005  manual   
354365  2016-03-14 17:48:27   2200          NaN              2005     NaN   
354366  2016-03-05 19:56:21   1199  convertible              2000    auto   
354367  2016-03-19 18:57:12   9200          bus              1996  manual   
354368  2016-03-20 19:41:08   3400        wagon              2002  manual   

        Power        Model  Kilometer  RegistrationMonth  FuelType  \
0           0         golf     150000                  0    petrol   
1         190          NaN     125000                  5  gasoline   
2         163        grand     125000                  8  gasoline   
3          75         golf     150000                  6    petrol   
4          69        fabia      90000                  7  gasoline   
...       ...          ...        ...                ...       ...   
354364      0         colt     150000                  7    petrol   
354365      0          NaN      20000                  1       NaN   
354366    101       fortwo     125000                  3    petrol   
354367    102  transporter     150000                  3  gasoline   
354368    100         golf     150000                  6  gasoline   

                 Brand NotRepaired          DateCreated  NumberOfPictures  \
0           volkswagen         NaN  2016-03-24 00:00:00                 0   
1                 audi         yes  2016-03-24 00:00:00                 0   
2                 jeep         NaN  2016-03-14 00:00:00                 0   
3           volkswagen          no  2016-03-17 00:00:00                 0   
4                skoda          no  2016-03-31 00:00:00                 0   
...                ...         ...                  ...               ...   
354364      mitsubishi         yes  2016-03-21 00:00:00                 0   
354365  sonstige_autos         NaN  2016-03-14 00:00:00                 0   
354366           smart          no  2016-03-05 00:00:00                 0   
354367      volkswagen          no  2016-03-19 00:00:00                 0   
354368      volkswagen         NaN  2016-03-20 00:00:00                 0   

        PostalCode             LastSeen  
0            70435  2016-04-07 03:16:57  
1            66954  2016-04-07 01:46:50  
2            90480  2016-04-05 12:47:46  
3            91074  2016-03-17 17:40:17  
4            60437  2016-04-06 10:17:21  
...            ...                  ...  
354364        2694  2016-03-21 10:42:49  
354365       39576  2016-04-06 00:46:52  
354366       26135  2016-03-11 18:17:12  
354367       87439  2016-04-07 07:15:26  
354368       40764  2016-03-24 12:45:21  

[354369 rows x 16 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Kilometer          354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

In [ ]:
df.isna().mean() * 100

DateCrawled           0.000000
Price                 0.000000
VehicleType          10.579368
RegistrationYear      0.000000
Gearbox               5.596709
Power                 0.000000
Model                 5.560588
Kilometer             0.000000
RegistrationMonth     0.000000
FuelType              9.282697
Brand                 0.000000
NotRepaired          20.079070
DateCreated           0.000000
NumberOfPictures      0.000000
PostalCode            0.000000
LastSeen              0.000000
dtype: float64

### Вывод

**В результате предварительного анализа установлено:**

1. В столбце VehicleType (тип автомобильного кузова) - 10.6 % пропусков. Тип данных - object.
2. В столбце Gearbox (тип коробки передач) - 5.6 % пропусков. Тип данных - object.
3. В столбце Model (модель автомобиля) - 5.6 % пропусков. Тип данных - object.
4. В столбце FuelType (тип топлива) - 9.3 % пропусков. Тип данных - object.
5. В столбце NotRepaired (была машина в ремонте или нет) - 20.1 % пропусков. Тип данных - object.

### Поиск дубликатов

In [ ]:
df.duplicated().sum()

4

In [ ]:
df = df.drop_duplicates().reset_index(drop=True) # удаление явных дубликатов (с удалением старых индексов и формированием новых)

In [ ]:
df.duplicated().sum() # проверка на отсутствие дубликатов

0

### Поиск неявных дубликатов

In [ ]:
#переведем столбецы с датой в формат datetime64[ns]
df['DateCrawled'] = pd.to_datetime(df['DateCrawled'], format = '%Y-%m-%d')
df['DateCreated'] = pd.to_datetime(df['DateCreated'], format = '%Y-%m-%d')
df['LastSeen'] = pd.to_datetime(df['LastSeen'], format = '%Y-%m-%d')

In [ ]:
df['DateCrawled'].value_counts(ascending=False)

2016-03-24 14:49:47    7
2016-03-19 21:49:56    6
2016-03-26 22:57:31    6
2016-03-11 15:36:59    5
2016-03-28 10:36:47    5
                      ..
2016-03-13 23:39:11    1
2016-03-20 16:48:06    1
2016-03-22 21:51:04    1
2016-03-11 16:25:36    1
2016-03-19 18:57:12    1
Name: DateCrawled, Length: 271174, dtype: int64

In [ ]:
df.query('DateCrawled == "2016-03-24 14:49:47"')

DateCrawled  Price  VehicleType  RegistrationYear Gearbox  \
33696  2016-03-24 14:49:47   3799        sedan              2002  manual   
41468  2016-03-24 14:49:47    990        wagon              2000  manual   
43209  2016-03-24 14:49:47   3000        sedan              2004  manual   
72521  2016-03-24 14:49:47   4800        small              1982  manual   
131819 2016-03-24 14:49:47    850  convertible              1992  manual   
150172 2016-03-24 14:49:47    399        small              1999  manual   
218109 2016-03-24 14:49:47   7500        sedan              2002    auto   

        Power     Model  Kilometer  RegistrationMonth  FuelType  \
33696     101        a4     150000                 10  gasoline   
41468       0    passat     150000                  3  gasoline   
43209     116       3er     150000                  1  gasoline   
72521      34    kaefer      70000                 12    petrol   
131819    198        80     150000                  5    petrol   
150172      0       NaN     150000                  0    petrol   
218109    309  e_klasse     150000                 11    petrol   

                Brand NotRepaired DateCreated  NumberOfPictures  PostalCode  \
33696            audi         yes  2016-03-24                 0       72800   
41468      volkswagen         NaN  2016-03-24                 0       83646   
43209             bmw          no  2016-03-24                 0       78467   
72521      volkswagen          no  2016-03-24                 0       14947   
131819           audi         yes  2016-03-24                 0       90443   
150172         lancia         yes  2016-03-24                 0       89231   
218109  mercedes_benz          no  2016-03-24                 0       50321   

                  LastSeen  
33696  2016-04-07 06:45:17  
41468  2016-04-01 23:41:53  
43209  2016-04-07 06:45:18  
72521  2016-04-07 06:45:17  
131819 2016-03-30 16:15:57  
150172 2016-04-07 06:45:19  
218109 2016-03-27 11:44:58

In [ ]:
df.query('DateCrawled == "2016-03-19 21:49:56"')

DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
71023  2016-03-19 21:49:56    899       small              2001  manual   
107543 2016-03-19 21:49:56   8250       wagon              2005    auto   
127851 2016-03-19 21:49:56   3299         bus              2002  manual   
146089 2016-03-19 21:49:56   3650         NaN              2017  manual   
192092 2016-03-19 21:49:56    700       small              1999  manual   
299887 2016-03-19 21:49:56   3100       small              2008  manual   

        Power        Model  Kilometer  RegistrationMonth  FuelType  \
71023      58        agila     150000                  6    petrol   
107543    224           a6     150000                 10  gasoline   
127851    116        other     125000                 11    petrol   
146089     77  transporter      20000                  0  gasoline   
192092     54        corsa     125000                  0    petrol   
299887     86        ibiza     150000                  4    petrol   

             Brand NotRepaired DateCreated  NumberOfPictures  PostalCode  \
71023         opel          no  2016-03-19                 0       55128   
107543        audi         NaN  2016-03-19                 0       65929   
127851     citroen          no  2016-03-19                 0       52249   
146089  volkswagen         NaN  2016-03-19                 0       35713   
192092        opel         NaN  2016-03-19                 0       35683   
299887        seat          no  2016-03-19                 0       97837   

                  LastSeen  
71023  2016-04-02 11:15:42  
107543 2016-04-04 08:45:52  
127851 2016-03-23 11:45:08  
146089 2016-03-25 09:47:36  
192092 2016-03-24 18:15:38  
299887 2016-03-19 22:43:16

In [ ]:
df.query('DateCrawled == "2016-03-26 22:57:31"')

DateCrawled  Price  VehicleType  RegistrationYear Gearbox  \
67475  2016-03-26 22:57:31  13000        small              2010    auto   
71500  2016-03-26 22:57:31   4150        sedan              1997    auto   
96762  2016-03-26 22:57:31    600          NaN              2000     NaN   
206202 2016-03-26 22:57:31    375          NaN              2017  manual   
280872 2016-03-26 22:57:31  11499        sedan              2009  manual   
282968 2016-03-26 22:57:31  11500  convertible              2008  manual   

        Power   Model  Kilometer  RegistrationMonth  FuelType       Brand  \
67475     105    polo      20000                 12    petrol  volkswagen   
71500     179     NaN     150000                 10    petrol        audi   
96762      90   bravo     150000                  0       NaN        fiat   
206202      0     3er     150000                  0    petrol         bmw   
280872    204     1er     125000                  6  gasoline         bmw   
282968    116  cooper      50000                  5    petrol        mini   

       NotRepaired DateCreated  NumberOfPictures  PostalCode  \
67475           no  2016-03-26                 0       91207   
71500           no  2016-03-26                 0        9128   
96762          NaN  2016-03-26                 0       59964   
206202         NaN  2016-03-26                 0       71636   
280872          no  2016-03-26                 0       36272   
282968          no  2016-03-26                 0       88048   

                  LastSeen  
67475  2016-04-07 02:46:52  
71500  2016-04-01 05:47:34  
96762  2016-03-27 08:41:12  
206202 2016-03-30 12:18:14  
280872 2016-04-02 16:15:57  
282968 2016-04-07 02:46:54

In [ ]:
df.query('DateCrawled == "2016-03-11 15:36:59"')

DateCrawled  Price  VehicleType  RegistrationYear Gearbox  \
48628  2016-03-11 15:36:59   2800        sedan              2000  manual   
113135 2016-03-11 15:36:59      0        wagon              1997  manual   
196075 2016-03-11 15:36:59    600          NaN              2016     NaN   
284579 2016-03-11 15:36:59   2600  convertible              2001    auto   
331423 2016-03-11 15:36:59    150        small              1993  manual   

        Power    Model  Kilometer  RegistrationMonth FuelType       Brand  \
48628     209    other     150000                  4   petrol        audi   
113135    125       a4     150000                  0   petrol        audi   
196075     75     golf     150000                  8      NaN  volkswagen   
284579    109  2_reihe     125000                  8   petrol     peugeot   
331423     50    corsa     150000                  3   petrol        opel   

       NotRepaired DateCreated  NumberOfPictures  PostalCode  \
48628           no  2016-03-11                 0       71067   
113135         yes  2016-03-11                 0       87730   
196075         NaN  2016-03-11                 0       45888   
284579          no  2016-03-11                 0       67071   
331423         NaN  2016-03-11                 0       12099   

                  LastSeen  
48628  2016-03-17 01:44:35  
113135 2016-04-05 11:46:38  
196075 2016-03-17 00:15:46  
284579 2016-03-18 00:18:02  
331423 2016-03-11 16:42:14

In [ ]:
df.query('DateCrawled == "2016-03-28 10:36:47"') 

DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
30031  2016-03-28 10:36:47   1200         NaN              2017  manual   
50486  2016-03-28 10:36:47  18800       wagon              2011    auto   
86793  2016-03-28 10:36:47    800       sedan              1998  manual   
187818 2016-03-28 10:36:47   1050       wagon              1998    auto   
190182 2016-03-28 10:36:47   1300       sedan              1998  manual   

        Power     Model  Kilometer  RegistrationMonth  FuelType  \
30031      75    kadett     150000                  8    petrol   
50486     170  e_klasse     150000                  6  gasoline   
86793     125        a3     150000                  2    petrol   
187818    101    vectra     150000                  5    petrol   
190182      0   octavia     150000                  1    petrol   

                Brand NotRepaired DateCreated  NumberOfPictures  PostalCode  \
30031            opel         NaN  2016-03-28                 0       93155   
50486   mercedes_benz          no  2016-03-28                 0        1326   
86793            audi          no  2016-03-28                 0       26721   
187818           opel          no  2016-03-28                 0       67365   
190182          skoda         NaN  2016-03-28                 0       27755   

                  LastSeen  
30031  2016-04-06 11:46:51  
50486  2016-04-06 11:46:51  
86793  2016-04-06 11:46:52  
187818 2016-04-06 13:44:51  
190182 2016-04-02 15:15:31

### Вывод

Среди первых пяти потенциальных неявных дубликатов в списке (df['DateCrawled'].value_counts(ascending=False)) - неявные дубликаты не обнаружены.

### Просмотр описательной статистики и поиск аномалий

In [ ]:
df.describe()

Price  RegistrationYear          Power      Kilometer  \
count  354365.000000     354365.000000  354365.000000  354365.000000   
mean     4416.679830       2004.234481     110.093816  128211.363989   
std      4514.176349         90.228466     189.851330   37905.083858   
min         0.000000       1000.000000       0.000000    5000.000000   
25%      1050.000000       1999.000000      69.000000  125000.000000   
50%      2700.000000       2003.000000     105.000000  150000.000000   
75%      6400.000000       2008.000000     143.000000  150000.000000   
max     20000.000000       9999.000000   20000.000000  150000.000000   

       RegistrationMonth  NumberOfPictures     PostalCode  
count      354365.000000          354365.0  354365.000000  
mean            5.714650               0.0   50508.503800  
std             3.726432               0.0   25783.100078  
min             0.000000               0.0    1067.000000  
25%             3.000000               0.0   30165.000000  
50%             6.000000               0.0   49413.000000  
75%             9.000000               0.0   71083.000000  
max            12.000000               0.0   99998.000000

In [ ]:
df['RegistrationYear'].value_counts(ascending=False)

2000    24490
1999    22727
2005    22109
2001    20123
2006    19900
        ...  
6500        1
2290        1
1001        1
9450        1
3500        1
Name: RegistrationYear, Length: 151, dtype: int64

In [ ]:
df['RegistrationYear'].sort_values().unique()

array([1000, 1001, 1039, 1111, 1200, 1234, 1253, 1255, 1300, 1400, 1500,
       1600, 1602, 1688, 1800, 1910, 1915, 1919, 1920, 1923, 1925, 1927,
       1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936, 1937, 1938,
       1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950,
       1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961,
       1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972,
       1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983,
       1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994,
       1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,
       2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016,
       2017, 2018, 2019, 2066, 2200, 2222, 2290, 2500, 2800, 2900, 3000,
       3200, 3500, 3700, 3800, 4000, 4100, 4500, 4800, 5000, 5300, 5555,
       5600, 5900, 5911, 6000, 6500, 7000, 7100, 7500, 7800, 8000, 8200,
       8455, 8500, 8888, 9000, 9229, 9450, 9996, 99

In [ ]:
bag_RegistrationYear = df.query('RegistrationYear == [1000, 1001, 1039, 1111, 1200, 1234, 1253, 1255, 1300, 1400, 1500, 1600, 1602, 1688, 1800, 2017, 2018, 2019, 2066, 2200, 2222, 2290, 2500, 2800, 2900, 3000, 3200, 3500, 3700, 3800, 4000, 4100, 4500, 4800, 5000, 5300, 5555, 5600, 5900, 5911, 6000, 6500, 7000, 7100, 7500, 7800, 8000, 8200, 8455, 8500, 8888, 9000, 9229, 9450, 9996, 9999]')

In [ ]:
len(bag_RegistrationYear)

14596

In [ ]:
len(bag_RegistrationYear) / len(df) * 100

4.118916935927645

Аномальных данных с не корректным годом порядка 4 %

Так как год регистрации автомобиля вряд ли был в позапрошлом тысячелетии, а также не может быть после даты скачивания анкеты из базы, то целесообразно убрать данные не входящие в промежутком с 1900 по 2016.

In [ ]:
df['Power'].sort_values().unique()

array([    0,     1,     2,     3,     4,     5,     6,     7,     8,
           9,    10,    11,    12,    13,    14,    15,    16,    17,
          18,    19,    20,    21,    22,    23,    24,    25,    26,
          27,    28,    29,    30,    31,    32,    33,    34,    35,
          36,    37,    38,    39,    40,    41,    42,    43,    44,
          45,    46,    47,    48,    49,    50,    51,    52,    53,
          54,    55,    56,    57,    58,    59,    60,    61,    62,
          63,    64,    65,    66,    67,    68,    69,    70,    71,
          72,    73,    74,    75,    76,    77,    78,    79,    80,
          81,    82,    83,    84,    85,    86,    87,    88,    89,
          90,    91,    92,    93,    94,    95,    96,    97,    98,
          99,   100,   101,   102,   103,   104,   105,   106,   107,
         108,   109,   110,   111,   112,   113,   114,   115,   116,
         117,   118,   119,   120,   121,   122,   123,   124,   125,
         126,   127,

Минимальная мощность автомобиля исходя из источника (https://avtika.ru/kakoy-minimalnyy-obem-dvigatelya-avtomobilya/) составляет порядка 52 л.с., а максимальная (https://rating-avto.ru/raznoe/samyie-moshhnyie-avtomobilnyie-dvigateli.html#:~:text=Bugatti%208.0%20W16&text=%D0%9C%D0%BE%D1%89%D0%BD%D0%BE%D1%81%D1%82%D1%8C%20%D1%82%D0%B0%D0%BA%D0%BE%D0%B3%D0%BE%20%D0%B0%D0%B2%D1%82%D0%BE%D0%BC%D0%BE%D0%B1%D0%B8%D0%BB%D1%8C%D0%BD%D0%BE%D0%B3%D0%BE%20%D0%BC%D0%BE%D1%82%D0%BE%D1%80%D0%B0%20%D0%B4%D0%BE%D1%81%D1%82%D0%B8%D0%B3%D0%B0%D0%B5%D1%82,%D0%BE%D1%82%D0%BB%D0%B8%D1%87%D0%B0%D0%B5%D1%82%D1%81%D1%8F%20%D0%BD%D0%B5%D0%B2%D0%B5%D1%80%D0%BE%D1%8F%D1%82%D0%BD%D0%BE%20%D0%B1%D0%BE%D0%BB%D1%8C%D1%88%D0%B8%D0%BC%20%D1%80%D0%B0%D1%81%D1%85%D0%BE%D0%B4%D0%BE%D0%BC%20%D1%82%D0%BE%D0%BF%D0%BB%D0%B8%D0%B2%D0%B0) порядка 1500 л.с.

In [ ]:
b = np.array([    0,     1,     2,     3,     4,     5,     6,     7,     8,
           9,    10,    11,    12,    13,    14,    15,    16,    17,
          18,    19,    20,    21,    22,    23,    24,    25,    26,
          27,    28,    29,    30,    31,    32,    33,    34,    35,
          36,    37,    38,    39,    40,    41,    42,    43,    44,
          45,    46,    47,    48,    49,    50,    51,  1501,  1502,  1503,  1506,  1521,  1548,  1595,  1596,  1597,
        1598,  1600,  1625,  1631,  1653,  1659,  1689,  1700,  1701,
        1703,  1704,  1707,  1753,  1771,  1779,  1780,  1781,  1783,
        1793,  1796,  1799,  1800,  1801,  1870,  1895,  1896,  1900,
        1910,  1920,  1922,  1933,  1937,  1968,  1986,  1988,  1992,
        1993,  1995,  1998,  1999,  2000,  2004,  2005,  2007,  2009,
        2016,  2017,  2018,  2172,  2200,  2201,  2331,  2340,  2389,
        2402,  2461,  2598,  2729,  2789,  2792,  2799,  3000,  3199,
        3454,  3500,  3750,  4400,  4507,  4700,  5000,  5411,  5420,
        5575,  5809,  5815,  5867,  6006,  6010,  6011,  6012,  6018,
        6045,  6062,  6226,  6512,  6920,  7508,  7511,  7512,  7515,
        7518,  7529,  7544,  8011,  8259,  8404,  8500,  9000,  9007,
        9010,  9011,  9012,  9013,  9710, 10000, 10110, 10218, 10311,
       10317, 10520, 10522, 10710, 10910, 10912, 11011, 11025, 11111,
       11509, 11530, 11635, 12012, 12510, 12512, 12684, 13616, 13636,
       14009, 15001, 15016, 15017, 15020, 15033, 16011, 16051, 16311,
       16312, 17011, 17019, 17410, 17700, 17932, 19208, 19211, 19312,
       20000])

In [ ]:
bag_Power_count = 0
for i in range(len(df)):
  if df['Power'][i] in b:
    bag_Power_count = bag_Power_count + 1


In [ ]:
bag_Power_count

51095

In [ ]:
bag_Power_count / len(df) * 100 

14.418749029954988

Аномальных данных с не корректный мощностью порядка 14 %

In [ ]:
df['Price'].value_counts(ascending=False)

0        10772
500       5670
1500      5394
1000      4648
1200      4594
         ...  
1368         1
233          1
11080        1
16340        1
10985        1
Name: Price, Length: 3731, dtype: int64

In [ ]:
bag_Price_count = 10772

In [ ]:
bag_Price_count / len(df) * 100

3.0398035923412303

## Вывод

**В результате подготовки данных установлено:**

1. В столбце VehicleType (тип автомобильного кузова) - 10.6 % пропусков. Тип данных - object.
2. В столбце Gearbox (тип коробки передач) - 5.6 % пропусков. Тип данных - object.
3. В столбце Model (модель автомобиля) - 5.6 % пропусков. Тип данных - object.
4. В столбце FuelType (тип топлива) - 9.3 % пропусков. Тип данных - object.
5. В столбце NotRepaired (была машина в ремонте или нет) - 20.1 % пропусков. Тип данных - object.
6. Среди первых пяти потенциальных неявных дубликатов в списке (df['DateCrawled'].value_counts(ascending=False)) - неявные дубликаты не обнаружены.
7. Тип данных либо 'int', либо 'object'.
8. Аномальных данных в столбце 'RegistrationYear' порядка 4 %.
9. Аномальных данных в столбце 'Power' порядка 14 %.
10. Также имеется не значительное количество аномальных данных в столбце 'Price'.

### Устранение аномалий и пропусков

In [ ]:
df.isna().mean() * 100

DateCrawled           0.000000
Price                 0.000000
VehicleType          10.579487
RegistrationYear      0.000000
Gearbox               5.596772
Power                 0.000000
Model                 5.560651
Kilometer             0.000000
RegistrationMonth     0.000000
FuelType              9.282802
Brand                 0.000000
NotRepaired          20.079297
DateCreated           0.000000
NumberOfPictures      0.000000
PostalCode            0.000000
LastSeen              0.000000
dtype: float64

Процент не корректных данных очень большой (аномалий и пропусков). Такие данные хотелось бы вернуть тому, кто их предоставил, НО МЫ ЛЮБИМ DS И БУДЕМ РАБОТАТЬ С ТЕМ, ЧТО ИМЕЕТСЯ!!!  В целом, данных достаточно много (354369) и основная задача создать модель, которая будет точно предсказывать стоимость (значение метрики RMSE должно быть меньше 2500). Обучающие данные должны быть максимально приближены к реальным, однако, при заполнении пропусков средними или медианными значениями это может привести к недостаточно точной работе модели, плюс наличие большого количества аномалий также пагубно скажется на качестве модели. Устранение пропусков и аномалий через удаление - это то, что не повлияет на истинность оставшихся значений по которым будет обучаться модель. Попробуем пойти именно по этому пути и, если удастся достичь значения метрики RMSE меньше 2500 - будем считать, что примененный способ борьбы с пропусками и аномалиями уместен.

In [ ]:
for i in range(len(df)):
  if df['Power'][i] in b:
    df['Power'][i] = np.NaN

In [ ]:
for j in range(len(df)):
  if df['Price'][j] < 1:
    df['Price'][j] = np.NaN

In [ ]:
df = df.query('RegistrationYear != [1000, 1001, 1039, 1111, 1200, 1234, 1253, 1255, 1300, 1400, 1500, 1600, 1602, 1688, 1800, 2017, 2018, 2019, 2066, 2200, 2222, 2290, 2500, 2800, 2900, 3000, 3200, 3500, 3700, 3800, 4000, 4100, 4500, 4800, 5000, 5300, 5555, 5600, 5900, 5911, 6000, 6500, 7000, 7100, 7500, 7800, 8000, 8200, 8455, 8500, 8888, 9000, 9229, 9450, 9996, 9999]')

In [ ]:
df

DateCrawled    Price  VehicleType  RegistrationYear Gearbox  \
0      2016-03-24 11:52:17    480.0          NaN              1993  manual   
1      2016-03-24 10:58:45  18300.0        coupe              2011  manual   
2      2016-03-14 12:52:21   9800.0          suv              2004    auto   
3      2016-03-17 16:54:04   1500.0        small              2001  manual   
4      2016-03-31 17:25:20   3600.0        small              2008  manual   
...                    ...      ...          ...               ...     ...   
354360 2016-03-21 09:50:58      NaN          NaN              2005  manual   
354361 2016-03-14 17:48:27   2200.0          NaN              2005     NaN   
354362 2016-03-05 19:56:21   1199.0  convertible              2000    auto   
354363 2016-03-19 18:57:12   9200.0          bus              1996  manual   
354364 2016-03-20 19:41:08   3400.0        wagon              2002  manual   

        Power        Model  Kilometer  RegistrationMonth  FuelType  \
0         NaN         golf     150000                  0    petrol   
1       190.0          NaN     125000                  5  gasoline   
2       163.0        grand     125000                  8  gasoline   
3        75.0         golf     150000                  6    petrol   
4        69.0        fabia      90000                  7  gasoline   
...       ...          ...        ...                ...       ...   
354360    NaN         colt     150000                  7    petrol   
354361    NaN          NaN      20000                  1       NaN   
354362  101.0       fortwo     125000                  3    petrol   
354363  102.0  transporter     150000                  3  gasoline   
354364  100.0         golf     150000                  6  gasoline   

                 Brand NotRepaired DateCreated  NumberOfPictures  PostalCode  \
0           volkswagen         NaN  2016-03-24                 0       70435   
1                 audi         yes  2016-03-24                 0       66954   
2                 jeep         NaN  2016-03-14                 0       90480   
3           volkswagen          no  2016-03-17                 0       91074   
4                skoda          no  2016-03-31                 0       60437   
...                ...         ...         ...               ...         ...   
354360      mitsubishi         yes  2016-03-21                 0        2694   
354361  sonstige_autos         NaN  2016-03-14                 0       39576   
354362           smart          no  2016-03-05                 0       26135   
354363      volkswagen          no  2016-03-19                 0       87439   
354364      volkswagen         NaN  2016-03-20                 0       40764   

                  LastSeen  
0      2016-04-07 03:16:57  
1      2016-04-07 01:46:50  
2      2016-04-05 12:47:46  
3      2016-03-17 17:40:17  
4      2016-04-06 10:17:21  
...                    ...  
354360 2016-03-21 10:42:49  
354361 2016-04-06 00:46:52  
354362 2016-03-11 18:17:12  
354363 2016-04-07 07:15:26  
354364 2016-03-24 12:45:21  

[339769 rows x 16 columns]

In [ ]:
df = df.dropna()
df = df.reset_index(drop=True)

### Разделение выборки на тестовую, тренировочную и валидационную, а также масштабирование и кодирование

In [ ]:
features_dumm = df.drop(columns=['DateCrawled', 'Price', 'DateCreated', 'LastSeen'])
features_Ord = df.drop(columns=['DateCrawled', 'Price', 'DateCreated', 'LastSeen'])

In [ ]:
target_dumm = df['Price']
target_Ord = df['Price']

In [ ]:
features_dumm

VehicleType  RegistrationYear Gearbox  Power        Model  Kilometer  \
0             small              2001  manual   75.0         golf     150000   
1             small              2008  manual   69.0        fabia      90000   
2             sedan              1995  manual  102.0          3er     150000   
3       convertible              2004  manual  109.0      2_reihe     150000   
4             sedan              2004  manual  105.0      3_reihe     150000   
...             ...               ...     ...    ...          ...        ...   
227074        sedan              2008  manual  105.0         leon     150000   
227075        sedan              2010  manual  140.0         golf     150000   
227076        sedan              2004  manual  225.0         leon     150000   
227077  convertible              2000    auto  101.0       fortwo     125000   
227078          bus              1996  manual  102.0  transporter     150000   

        RegistrationMonth  FuelType       Brand NotRepaired  NumberOfPictures  \
0                       6    petrol  volkswagen          no                 0   
1                       7  gasoline       skoda          no                 0   
2                      10    petrol         bmw         yes                 0   
3                       8    petrol     peugeot          no                 0   
4                      12    petrol       mazda          no                 0   
...                   ...       ...         ...         ...               ...   
227074                  7  gasoline        seat          no                 0   
227075                  7  gasoline  volkswagen          no                 0   
227076                  5    petrol        seat         yes                 0   
227077                  3    petrol       smart          no                 0   
227078                  3  gasoline  volkswagen          no                 0   

        PostalCode  
0            91074  
1            60437  
2            33775  
3            67112  
4            96224  
...            ...  
227074       45896  
227075       75223  
227076       96465  
227077       26135  
227078       87439  

[227079 rows x 12 columns]

In [ ]:
features_dumm.shape

(227079, 12)

In [ ]:
target_dumm.shape

(227079,)

In [ ]:
# прямое кодирование
features_dumm = pd.get_dummies(features_dumm, drop_first=True)

In [ ]:
# порядковое кодирование
ord_enc = OrdinalEncoder()

In [ ]:
features_Ord[['VehicleType', 'Gearbox', 'Model', 'FuelType', 'Brand', 'NotRepaired']] = ord_enc.fit_transform(features_Ord[['VehicleType', 'Gearbox', 'Model', 'FuelType', 'Brand', 'NotRepaired']])

In [ ]:
# деление на обучающую и валидационно-тестовую выборку (60:40)
features_train_dumm, features_valid_test_dumm, target_train_dumm, target_valid_test_dumm = train_test_split(
    features_dumm, target_dumm, test_size=0.4, random_state=12345)

In [ ]:
# деление валидационно-тестовой выборки (50:50)
features_valid_dumm, features_test_dumm, target_valid_dumm, target_test_dumm = train_test_split(
    features_valid_test_dumm, target_valid_test_dumm, test_size=0.5, random_state=12345)

In [ ]:
# деление на обучающую и валидационно-тестовую выборку (60:40)
features_train_Ord, features_valid_test_Ord, target_train_Ord, target_valid_test_Ord = train_test_split(
    features_Ord, target_Ord, test_size=0.4, random_state=12345)

In [ ]:
# деление валидационно-тестовой выборки (50:50)
features_valid_Ord, features_test_Ord, target_valid_Ord, target_test_Ord = train_test_split(
    features_valid_test_Ord, target_valid_test_Ord, test_size=0.5, random_state=12345)

In [ ]:
# масштабирование признаков
numeric = ['RegistrationYear', 'Power', 'Kilometer', 'RegistrationMonth', 'NumberOfPictures','PostalCode']

scaler = StandardScaler()
scaler.fit(features_dumm[numeric])
features_dumm[numeric] = scaler.transform(features_dumm[numeric])

In [ ]:
# масштабирование признаков
numeric = ['RegistrationYear', 'Power', 'Kilometer', 'RegistrationMonth', 'NumberOfPictures','PostalCode']

scaler = StandardScaler()
scaler.fit(features_Ord[numeric])
features_Ord[numeric] = scaler.transform(features_Ord[numeric])

In [ ]:
features_dumm.head(3)

RegistrationYear     Power  Kilometer  RegistrationMonth  NumberOfPictures  \
0         -0.386434 -0.908448   0.619292          -0.065857               0.0   
1          0.802918 -1.016750  -1.000968           0.224137               0.0   
2         -1.405879 -0.421085   0.619292           1.094121               0.0   

   PostalCode  VehicleType_convertible  VehicleType_coupe  VehicleType_other  \
0    1.516718                        0                  0                  0   
1    0.330551                        0                  0                  0   
2   -0.701717                        0                  0                  0   

   VehicleType_sedan  ...  Brand_saab  Brand_seat  Brand_skoda  Brand_smart  \
0                  0  ...           0           0            0            0   
1                  0  ...           0           0            1            0   
2                  1  ...           0           0            0            0   

   Brand_subaru  Brand_suzuki  Brand_toyota  Brand_volkswagen  Brand_volvo  \
0             0             0             0                 1            0   
1             0             0             0                 0            0   
2             0             0             0                 0            0   

   NotRepaired_yes  
0                0  
1                0  
2                1  

[3 rows x 305 columns]

In [ ]:
features_Ord.head(3)

VehicleType  RegistrationYear  Gearbox     Power  Model  Kilometer  \
0          5.0         -0.386434      1.0 -0.908448  115.0   0.619292   
1          5.0          0.802918      1.0 -1.016750  100.0  -1.000968   
2          4.0         -1.405879      1.0 -0.421085   11.0   0.619292   

   RegistrationMonth  FuelType  Brand  NotRepaired  NumberOfPictures  \
0          -0.065857       6.0   36.0          0.0               0.0   
1           0.224137       2.0   31.0          0.0               0.0   
2           1.094121       6.0    2.0          1.0               0.0   

   PostalCode  
0    1.516718  
1    0.330551  
2   -0.701717

In [ ]:
features_Ord.shape

(227079, 12)

In [ ]:
features_dumm.shape

(227079, 305)

## Обучение моделей

### LinearRegression

In [ ]:
# на данных с прямым кодированием
%%time
# расчет для LinearRegression
dumm_model_LinearRegression = LinearRegression()
dumm_model_LinearRegression.fit(features_train_dumm, target_train_dumm)
y_true = target_valid_dumm
y_pred = pd.DataFrame(dumm_model_LinearRegression.predict(features_valid_dumm))

result = mean_squared_error(y_true, y_pred)**0.5
result_1 = r2_score(y_true, y_pred)

print("Модель:", dumm_model_LinearRegression)
print("RMSE:", round(result, 1))
print("R2:", round(result_1, 3))

Модель: LinearRegression()
RMSE: 2566.0
R2: 0.71
CPU times: user 7.33 s, sys: 672 ms, total: 8 s
Wall time: 4.52 s


In [ ]:
# на данных с порядковым кодированием
%%time
# расчет для LinearRegression
model_LinearRegression = LinearRegression()
model_LinearRegression.fit(features_train_Ord, target_train_Ord)
y_true = target_valid_Ord
y_pred = pd.DataFrame(model_LinearRegression.predict(features_valid_Ord))

result = mean_squared_error(y_true, y_pred)**0.5
result_1 = r2_score(y_true, y_pred)

print("Модель:", model_LinearRegression)
print("RMSE:", round(result, 1))
print("R2:", round(result_1, 3))

Модель: LinearRegression()
RMSE: 2954.3
R2: 0.615
CPU times: user 89.1 ms, sys: 69.4 ms, total: 158 ms
Wall time: 83.1 ms


### RandomForestRegressor

In [ ]:
# на данных с прямым кодированием
%%time
# подбор гиперпараметров: глубины дерева, числа деревьев на основе RMSE
dumm_best_model_RandomForestRegressor = None
best_result = 100000000
best_result_1 = 0
for depth in range(1, 6, 1):
  for est in range(1, 6, 1):
    model = RandomForestRegressor(n_estimators=est, max_depth=depth, random_state=12345)
    model.fit(features_train_dumm, target_train_dumm)
    
    y_true = target_valid_dumm
    y_pred = pd.DataFrame(model.predict(features_valid_dumm))

    result = mean_squared_error(y_true, y_pred)**0.5
    result_1 = r2_score(y_true, y_pred)
    if result < best_result:
      dumm_best_model_RandomForestRegressor = model
      best_result = result
      best_result_1 = result_1

print("Лучшие параметры модели:", dumm_best_model_RandomForestRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345)
RMSE: 2375.6
R2: 0.751
CPU times: user 46.4 s, sys: 209 ms, total: 46.6 s
Wall time: 46.4 s


In [ ]:
# на данных с порядковым кодированием
%%time
# подбор гиперпараметров: глубины дерева, числа деревьев на основе RMSE
best_model_RandomForestRegressor = None
best_result = 100000000
best_result_1 = 0
for depth in range(1, 6, 1):
  for est in range(1, 6, 1):
    model = RandomForestRegressor(n_estimators=est, max_depth=depth, random_state=12345)
    model.fit(features_train_Ord, target_train_Ord)
    
    y_true = target_valid_Ord
    y_pred = pd.DataFrame(model.predict(features_valid_Ord))

    result = mean_squared_error(y_true, y_pred)**0.5
    result_1 = r2_score(y_true, y_pred)
    if result < best_result:
      best_model_RandomForestRegressor = model
      best_result = result
      best_result_1 = result_1

print("Лучшие параметры модели:", best_model_RandomForestRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345)
RMSE: 2379.1
R2: 0.75
CPU times: user 9.3 s, sys: 36.1 ms, total: 9.34 s
Wall time: 9.34 s


### DecisionTreeRegressor

In [ ]:
# на данных с прямым кодированием
%%time
# подбор гиперпараметров: глубины дерева, минимального числа samples_split на основе RMSE
dumm_best_model_DecisionTreeRegressor = None
best_result = 100000000
best_result_1 = 0
for depth in range(1, 6, 1):
  for spl in range(2, 7, 1):
    model = DecisionTreeRegressor(max_depth=depth, min_samples_split=spl, random_state=12345)
    model.fit(features_train_dumm, target_train_dumm)
    
    y_true = target_valid_dumm
    y_pred = pd.DataFrame(model.predict(features_valid_dumm))

    result = mean_squared_error(y_true, y_pred)**0.5
    result_1 = r2_score(y_true, y_pred)
    if result < best_result:
      dumm_best_model_DecisionTreeRegressor = model
      best_result = result
      best_result_1 = result_1

print("Лучшие параметры модели:", dumm_best_model_DecisionTreeRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: DecisionTreeRegressor(max_depth=5, random_state=12345)
RMSE: 2411.9
R2: 0.743
CPU times: user 24 s, sys: 110 ms, total: 24.1 s
Wall time: 25 s


In [ ]:
# на данных с порядковым кодированием
%%time
# подбор гиперпараметров: глубины дерева, минимального числа samples_split на основе RMSE
best_model_DecisionTreeRegressor = None
best_result = 100000000
best_result_1 = 0
for depth in range(1, 6, 1):
  for spl in range(2, 7, 1):
    model = DecisionTreeRegressor(max_depth=depth, min_samples_split=spl, random_state=12345)
    model.fit(features_train_Ord, target_train_Ord)
    
    y_true = target_valid_Ord
    y_pred = pd.DataFrame(model.predict(features_valid_Ord))

    result = mean_squared_error(y_true, y_pred)**0.5
    result_1 = r2_score(y_true, y_pred)
    if result < best_result:
      best_model_DecisionTreeRegressor = model
      best_result = result
      best_result_1 = result_1

print("Лучшие параметры модели:", best_model_DecisionTreeRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: DecisionTreeRegressor(max_depth=5, random_state=12345)
RMSE: 2416.4
R2: 0.742
CPU times: user 5.2 s, sys: 63.7 ms, total: 5.27 s
Wall time: 5.35 s


### DummyRegressor

In [ ]:
# на данных с прямым кодированием
%%time
# подбор гиперпараметров: ['mean', 'median'] на основе RMSE
dumm_best_model_DummyRegressor = None
best_result = 100000000
best_result_1 = 0
list_strategy = ['mean', 'median']
for strat in list_strategy:
  model = DummyRegressor(strategy=strat)
  model.fit(features_train_dumm, target_train_dumm)

  y_true = target_valid_dumm
  y_pred = pd.DataFrame(model.predict(features_valid_dumm))
  
  result = mean_squared_error(y_true, y_pred)**0.5
  result_1 = r2_score(y_true, y_pred)
  if result < best_result:
    dumm_best_model_DummyRegressor = model
    best_result = result
    best_result_1 = result_1

print("Лучшие параметры модели:", dumm_best_model_DummyRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: DummyRegressor()
RMSE: 4761.5
R2: -0.113
CPU times: user 15.5 ms, sys: 0 ns, total: 15.5 ms
Wall time: 14.5 ms


In [ ]:
# на данных с порядковым кодированием
%%time
# подбор гиперпараметров: ['mean', 'median'] на основе RMSE
best_model_DummyRegressor = None
best_result = 100000000
best_result_1 = 0
list_strategy = ['mean', 'median']
for strat in list_strategy:
  model = DummyRegressor(strategy=strat)
  model.fit(features_train_Ord, target_train_Ord)

  y_true = target_valid_Ord
  y_pred = pd.DataFrame(model.predict(features_valid_Ord))
  
  result = mean_squared_error(y_true, y_pred)**0.5
  result_1 = r2_score(y_true, y_pred)
  if result < best_result:
    best_model_DummyRegressor = model
    best_result = result
    best_result_1 = result_1

print("Лучшие параметры модели:", best_model_DummyRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: DummyRegressor()
RMSE: 4761.5
R2: -0.113
CPU times: user 23.8 ms, sys: 2.01 ms, total: 25.8 ms
Wall time: 23.7 ms


### LGBMRegressor

In [ ]:
# на данных с прямым кодированием
%%time
# подбор гиперпараметров: глубины дерева, числа деревьев на основе RMSE
dumm_best_model_LGBMRegressor = None
best_result = 100000000
best_result_1 = 0
for depth in range(1, 6, 1):
  for est in range(1, 6, 1):
    model = LGBMRegressor(n_estimators=est, max_depth=depth, random_state=12345)
    model.fit(features_train_dumm, target_train_dumm)
    
    y_true = target_valid_dumm
    y_pred = pd.DataFrame(model.predict(features_valid_dumm))

    result = mean_squared_error(y_true, y_pred)**0.5
    result_1 = r2_score(y_true, y_pred)
    if result < best_result:
      dumm_best_model_LGBMRegressor = model
      best_result = result
      best_result_1 = result_1

print("Лучшие параметры модели:", dumm_best_model_LGBMRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: LGBMRegressor(max_depth=5, n_estimators=5, random_state=12345)
RMSE: 3399.2
R2: 0.49
CPU times: user 40.5 s, sys: 1.78 s, total: 42.3 s
Wall time: 29.5 s


In [ ]:
# на данных с порядковым кодированием
%%time
# подбор гиперпараметров: глубины дерева, числа деревьев на основе RMSE
best_model_LGBMRegressor = None
best_result = 100000000
best_result_1 = 0
for depth in range(1, 6, 1):
  for est in range(1, 6, 1):
    model = LGBMRegressor(n_estimators=est, max_depth=depth, random_state=12345)
    model.fit(features_train_Ord, target_train_Ord)
    
    y_true = target_valid_Ord
    y_pred = pd.DataFrame(model.predict(features_valid_Ord))

    result = mean_squared_error(y_true, y_pred)**0.5
    result_1 = r2_score(y_true, y_pred)
    if result < best_result:
      best_model_LGBMRegressor = model
      best_result = result
      best_result_1 = result_1

print("Лучшие параметры модели:", best_model_LGBMRegressor)
print("RMSE:", round(best_result, 1))
print("R2:", round(result_1, 3))

Лучшие параметры модели: LGBMRegressor(max_depth=5, n_estimators=5, random_state=12345)
RMSE: 3400.7
R2: 0.49
CPU times: user 10.2 s, sys: 88.3 ms, total: 10.3 s
Wall time: 5.63 s


## Вывод

**В результате обучение моделей на данных с прямым кодированием и порядковым соответственно, установлено:**

1. Модель: LinearRegression(). RMSE: 2566. CPU times: user около 7 s. Модель: LinearRegression(). RMSE: 2954. CPU times: user около 89 ms.
2. Лучшие параметры модели: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 2376. CPU times: user около 46 s. Лучшие параметры модели: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 2379. CPU times: user около 9 s.
3. Лучшие параметры модели: DecisionTreeRegressor(max_depth=5, random_state=12345). RMSE: 2412. CPU times: user около 24 s. Лучшие параметры модели: DecisionTreeRegressor(max_depth=5, random_state=12345). RMSE: 2416. CPU times: user около 5 s.
4. Модель: DummyRegressor() RMSE: 4762. CPU times: user около 16 ms. Модель: DummyRegressor() RMSE: RMSE: 4762. CPU times: user около 24 ms.
5. Лучшие параметры модели: LGBMRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 3399. CPU times: user около 41 s. Лучшие параметры модели: LGBMRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 3401. CPU times: user около 10 s.


## Анализ моделей

In [ ]:
dumm_list_model = [dumm_model_LinearRegression, dumm_best_model_RandomForestRegressor, dumm_best_model_DecisionTreeRegressor, dumm_best_model_DummyRegressor, dumm_best_model_LGBMRegressor]

In [ ]:
# поиск наилучшей модели на ТЕСТОВОЙ выборке
best_model = None
best_result = 100000000
for mod in dumm_list_model:
  test_predictions = mod.predict(features_test_dumm)
  y_true = target_test_dumm
  y_pred = pd.DataFrame(mod.predict(features_test_dumm))
  result = mean_squared_error(y_true, y_pred)**0.5
  if result < best_result:
    best_result = result
    best_model = mod

print("Наилучшая модель на ТЕСТОВОЙ выборке:", best_model)
print("RMSE наилучшей модели =", round(best_result, 1))

Наилучшая модель на ТЕСТОВОЙ выборке: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345)
RMSE наилучшей модели = 2336.4


In [ ]:
list_model = [model_LinearRegression, best_model_RandomForestRegressor, best_model_DecisionTreeRegressor, best_model_DummyRegressor, best_model_LGBMRegressor]

In [ ]:
# поиск наилучшей модели на ТЕСТОВОЙ выборке
best_model = None
best_result = 100000000
for mod in list_model:
  test_predictions = mod.predict(features_test_Ord)
  y_true = target_test_Ord
  y_pred = pd.DataFrame(mod.predict(features_test_Ord))
  result = mean_squared_error(y_true, y_pred)**0.5
  if result < best_result:
    best_result = result
    best_model = mod

print("Наилучшая модель на ТЕСТОВОЙ выборке:", best_model)
print("RMSE наилучшей модели =", round(best_result, 1))

Наилучшая модель на ТЕСТОВОЙ выборке: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345)
RMSE наилучшей модели = 2340.1


# Общий вывод

**I. В результате подготовки данных установлено:**

1. В столбце VehicleType (тип автомобильного кузова) - 10.6 % пропусков. Тип данных - object.
2. В столбце Gearbox (тип коробки передач) - 5.6 % пропусков. Тип данных - object.
3. В столбце Model (модель автомобиля) - 5.6 % пропусков. Тип данных - object.
4. В столбце FuelType (тип топлива) - 9.3 % пропусков. Тип данных - object.
5. В столбце NotRepaired (была машина в ремонте или нет) - 20.1 % пропусков. Тип данных - object.
6. Среди первых пяти потенциальных неявных дубликатов в списке (df['DateCrawled'].value_counts(ascending=False)) - неявные дубликаты не обнаружены.
7. Тип данных либо 'int', либо 'object'.
8. Аномальных данных в столбце 'RegistrationYear' порядка 4 %.
9. Аномальных данных в столбце 'Power' порядка 14 %.
10. Также имеется не значительное количество аномальных данных в столбце 'Price'.

**II. Процент не корректных данных очень большой (аномалий и пропусков). В целом, данных достаточно много (354369) и основная задача создать модель, которая будет точно предсказывать стоимость (значение метрики RMSE должно быть меньше 2500). Обучающие данные должны быть максимально приближены к реальным, однако, при заполнении пропусков средними или медианными значениями это может привести к недостаточно точной работе модели, плюс наличие большого количества аномалий также пагубно скажется на качестве модели. Устранение пропусков и аномалий через удаление - это то, что не повлияет на истинность оставшихся значений по которым будет обучаться модель. Попробуем пойти именно по этому пути и, если удастся достичь значения метрики RMSE меньше 2500 - будем считать, что примененный способ борьбы с пропусками и аномалиями уместен.**

**III. В результате обучение моделейна на данных с прямым кодированием и порядковым соответственно, установлено:**

1. Модель: LinearRegression(). RMSE: 2566. CPU times: user около 7 s. Модель: LinearRegression(). RMSE: 2954. CPU times: user около 89 ms.
2. Лучшие параметры модели: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 2376. CPU times: user около 46 s. Лучшие параметры модели: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 2379. CPU times: user около 9 s.
3. Лучшие параметры модели: DecisionTreeRegressor(max_depth=5, random_state=12345). RMSE: 2412. CPU times: user около 24 s. Лучшие параметры модели: DecisionTreeRegressor(max_depth=5, random_state=12345). RMSE: 2416. CPU times: user около 5 s.
4. Модель: DummyRegressor() RMSE: 4762. CPU times: user около 16 ms. Модель: DummyRegressor() RMSE: RMSE: 4762. CPU times: user около 24 ms.
5. Лучшие параметры модели: LGBMRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 3399. CPU times: user около 41 s. Лучшие параметры модели: LGBMRegressor(max_depth=5, n_estimators=5, random_state=12345). RMSE: 3401. CPU times: user около 10 s.


**IV. В результате анализа моделей, установлено:**

1. Наилучшая модель на ТЕСТОВОЙ выборке: RandomForestRegressor(max_depth=5, n_estimators=5, random_state=12345).
2. RMSE наилучшей модели на данных с прямым кодированием = 2336.4.
3. RMSE наилучшей модели на данных с порядковым кодированием = 2340.1.